File: python/Q297_serialize_and_deserialize_binary_tree.py
#
# @lc app=leetcode id=297 lang=python3
#
# [297] Serialize and Deserialize Binary Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x: int) -> None:
        self.val: int = x
        self.left: TreeNode | None = None
        self.right: TreeNode | None = None


# @lc code=start
class Codec:
    def preorder(self, root: TreeNode | None, node_list: list[str]) -> None:
        if not root:
            node_list.append("N")
            return

        node_list.append(str(root.val))
        self.preorder(root.left, node_list)
        self.preorder(root.right, node_list)

    def serialize(self, root) -> str:
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        result: list[str] = []
        self.preorder(root, result)

        return ",".join(result)

    def create(self) -> TreeNode | None:
        value: str = self.node_list[self.index]
        if value == "N":
            self.index += 1
            return None

        new_node: TreeNode = TreeNode(int(value))
        self.index += 1

        new_node.left = self.create()
        new_node.right = self.create()

        return new_node

    def deserialize(self, data) -> TreeNode | None:
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        self.node_list: list[str] = data.split(",")
        self.index: int = 0
        return self.create()


# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
# @lc code=end

File: rust/src/leetcode/Q297_serialize_and_deserialize_binary_tree.rs
/*
 * @lc app=leetcode id=297 lang=rust
 *
 * [297] Serialize and Deserialize Binary Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

struct Codec {}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Codec {
    fn new() -> Self {
        Self {}
    }

    fn preorder_traversal(root: Option<Rc<RefCell<TreeNode>>>, nodes: &mut Vec<String>) {
        if let Some(node) = root {
            nodes.push(node.borrow().val.to_string());

            Self::preorder_traversal(node.borrow().left.clone(), nodes);
            Self::preorder_traversal(node.borrow().right.clone(), nodes);
        } else {
            nodes.push(String::from("N"));
        }
    }

    fn serialize(&self, root: Option<Rc<RefCell<TreeNode>>>) -> String {
        let mut nodes: Vec<String> = vec![];
        Self::preorder_traversal(root, &mut nodes);

        nodes.join(",")
    }

    fn build_node(nodes: &Vec<&str>, index: &mut usize) -> Option<Rc<RefCell<TreeNode>>> {
        *index += 1;

        if nodes[*index - 1] == "N" {
            return None;
        }

        let value: i32 = nodes[*index - 1].parse().unwrap();

        let node: Rc<RefCell<TreeNode>> = Rc::new(RefCell::new(TreeNode::new(value)));

        node.borrow_mut().left = Self::build_node(nodes, index);
        node.borrow_mut().right = Self::build_node(nodes, index);

        Some(node)
    }

    fn deserialize(&self, data: String) -> Option<Rc<RefCell<TreeNode>>> {
        let nodes: Vec<&str> = data.split(",").collect();
        let mut index: usize = 0;

        Self::build_node(&nodes, &mut index)
    }
}

// /**
//  * Your Codec object will be instantiated and called as such:
//  * let obj = Codec::new();
//  * let data: String = obj.serialize(strs);
//  * let ans: Option<Rc<RefCell<TreeNode>>> = obj.deserialize(data);
//  */
// @lc code=end
